src/schedule/zephyr_ll.c

```diff
@@ -0,0 +1,296 @@
+	/* check if task is already scheduled */
+	list_for_item(list, &sch->tasks) {
+		task_iter = container_of(list, struct task, list);
+
+		/* keep original start */
+		if (task_iter == task) {
+			spin_unlock(&sch->lock);
+			return 0;
+		}
+	}
+
+	pdata = task->priv_data;
+	pdata->period = period;
+
+	task->start = domain_time(sch) + delay;
+	if (sch->ll_domain->next_tick != UINT64_MAX &&
+	    sch->ll_domain->next_tick > task->start)
+		task->start = sch->ll_domain->next_tick;
+	task->state = SOF_TASK_STATE_PENDING;
+
+	/*
+	 * Tasks are added into the list from highest to lowest priority. This
+	 * way they can then be run in the same order. Tasks with the same
+	 * priority are served on a first-come-first-serve basis
+	 */
+	list_for_item(list, &sch->tasks) {
+		task_iter = container_of(list, struct task, list);
+		if (task->priority < task_iter->priority) {
+			list_item_append(&task->list, &task_iter->list);
+			break;
+		}
+	}
+
+	/*
+	 * If the task has not been added, it means that it has the lowest
+	 * priority and should be added at the end of the list
+	 */
+	if (list == &sch->tasks)
+		list_item_append(&task->list, &sch->tasks);
+
+	spin_unlock(&sch->lock);
+
+	ret = domain_register(sch->ll_domain, task, &zephyr_ll_run, sch);
+	if (ret < 0)
+		tr_err(&ll_tr, "zephyr_ll_task_schedule: cannot register domain %d",
+		       ret);
+
+	return 0;
+}
+
+/* This is synchronous - after this returns the object can be destroyed! */
+static int zephyr_ll_task_free(void *data, struct task *task)
+{
+	struct zephyr_ll *sch = data;
+
+	switch (task->state) {
+	case SOF_TASK_STATE_INIT:
+	case SOF_TASK_STATE_CANCEL:
+	case SOF_TASK_STATE_FREE:
+		break;
+	default:
+		spin_lock(&sch->lock);
+		/* Now the task is guaranteed to not be running */
+		list_item_del(&task->list);
+		spin_unlock(&sch->lock);
+	}
+
+	rfree(task->priv_data);
+
+	return 0;
+}
+
+/* This seems to be asynchronous? */
+static int zephyr_ll_task_cancel(void *data, struct task *task)
+{
+	/*
+	 * FIXME: this can be called from an interrupt, also on a different core
+	 * Once we have moved running tasks to a different list, only the status
+	 * check there will be protected by the spin-lock. Then we can take the
+	 * spin-lock here too. Doing this now can potentially deadlock.
+	 */
+	task->state = SOF_TASK_STATE_CANCEL;
+
+	return 0;
+}
+
+/*
+ * FIXME: clarify rescheduling semantics. Currently only DMA trace is using it.
+ * The goal is to reschedule a periodic task with a relatively long period to
+ * run at an earlier time. But the current ll-scheduler based implementation
+ * doesn't even attempt to check whether the task is in fact scheduled to run at
+ * an earlier time.
+ * While our implementation runs all tasks under a spin-lock we could do various
+ * checks, but when we switch over to drop the spin-lock when running tasks,
+ * we'll have to come up with a robust implementation.
+ */
+static int zephyr_ll_task_reschedule(void *data, struct task *task, uint64_t start)
```

Comment:
ack, lets just say -ENOTSUP here as DMA trace will be using a thread with Zephyr.